### ISSUE TITLE
RFC: Allow passphrased keys to be specified to pycloudlib

### ISSUE THREAD

**Issue by OddBloke** [canonical/pycloudlib#82]
TL;DR: Scroll down to "I think there are three paths forward".

Currently, pycloudlib's default behaviour is to use `~/.ssh/id_rsa` as the SSH key that is provided to, and used to access, all machines under test, with one exception: pycloudlib assumes that a key named with the current username exists in EC2, and will not create it if it is absent. pycloudlib also explicitly checks that the given key does not have a passphrase on it: it raises an exception before we even attempt to connect to the remote SSH host if it does.

This poses a problem for me: on my primary system, I use a non-default key for my work SSH access (`~/.ssh/id_rsa.canonical`) and I have it protected (at rest) with a passphrase. I spent some time yesterday trying to implement support for my usecase in pycloudlib with limited success; this issue captures my findings/thinking.

paramiko can actually use my passphrased key without issues, if I remove the guard we have implemented. I run an SSH agent and populate it with my key at login, so I don't have to type the passphrase every time I use the key; paramiko's [`connect` docs](http://docs.paramiko.org/en/stable/api/client.html#paramiko.client.SSHClient.connect) states they will attempt auth with "Any key we can find through an SSH agent", after attempting the specified key. The first use of my key fails (because there's no passphrase to unlock it), and then it is successfully used when paramiko falls back to the SSH agent.

My initial implementation path was to completely rely on paramiko's [default connection behaviour](http://docs.paramiko.org/en/stable/api/client.html#paramiko.client.SSHClient.connect): it will use an SSH agent if available, falling back to looking at `~/.ssh/id_rsa`, which closely approximates our current behaviour. However, this runs into issues: for most clouds we need to provide the public key file's contents to the cloud platform on launch: if we don't have any reference to the key (because we're relying on our SSH agent), then we can't readily do that. (paramiko _does_ have an API for querying the SSH agent, but it's not clear how you would select the correct key without already having either the public key material or its filename; in either case, there's then no need to use the SSH agent.) So this isn't the way to go.

I think there are three paths forward:

1. continue to disallow the use of passphrased keys (i.e. change nothing). I don't consider this viable, we should not be forcing our users to create and/or use less secure SSH keys.
2. don't explicitly disallow the use of passphrased keys (i.e. remove the guard from `_ssh_connect`). This would enable my use-case (via agent fallback), but would mean that use of passphrased keys in the absence of an agent will lead to a ten minute retry loop before failing (vs. the immediate failure we have now).
3. handle the SSH key lifecycle entirely within pycloudlib: instead of using the user's existing keys at all, pycloudlib could generate SSH keys for each instance it launches, cleaning them up after use. These single-use keys don't need the protection that a user's primary keys do, so we could guarantee they have no passphrase.

I think that looking at cost/reward, (2) is probably our best path forward in the short/medium term. We can ensure that none of the keys used for automated test running have passphrases, and users for whom this would change behaviour (i.e. those with passphrased default keys) will (a) already be running into issues with the current codebase, and (b) be present to observe the retry looping and therefore able to intervene.

(Probably better than removing the guard is to instead have it emit some sort of warning, something like: "Specified key requires a passphrase; it will need to be added to a running ssh-agent for tests to run successfully".)

**Comment by TheRealFalcon** [comment]
One other option I can think of is allow the user to specify a keypair, with the default being ~/.ssh/id_rsa. This is a little more flexible but puts some more burden on the user to have a separate, non-password protected keypair that can be used for cloud testing.

Option 2 is probably the better option though. Either way we should probably do it.

I haven't looked into the implementation, but is there a way we can make that 10 minute timeout shorter? Even globally, I don't see a reason why we shouldn't timeout after something like 30 seconds of waiting for an ssh connection.

**Comment by OddBloke** [comment]
On Wed, Jan 06, 2021 at 09:18:27AM -0800, James Falcon wrote:
> One other option I can think of is allow the user to specify a
> keypair, with the default being ~/.ssh/id_rsa. This is a little more
> flexible but puts some more burden on the user to have a separate,
> non-password protected keypair that can be used for cloud testing.

pycloudlib does already support using a non-default key (with
`BaseCloud.use_key`); we'll need to hook that up in cloud-init's
integration testing.  As far as pycloudlib is concerned, this is really
about how we treat passphrased keys.

> Option 2 is probably the better option though. Either way we should
> probably do it.

Cool, I'll start looking at making the changes properly.

> I haven't looked into the implementation, but is there a way we can
> make that 10 minute timeout shorter? Even globally, I don't see a
> reason why we shouldn't timeout after something like 30 seconds of
> waiting for an ssh connection.

Yeah, it's just looping 60 times, so we could drop that loop count down.
I'd be inclined to leave it as-is, however: some clouds really do take a
long time to bring up SSH, and if we're stuck in the retry loop then all
our tests are going to fail anyway (because they all share the same
unusable key), so there's nothing much lost by manually killing the test
run early to avoid the long timeout(s).  What do you think?

**Comment by TheRealFalcon** [comment]
Ah, right. That's fine. I think ideally, we'd have two separate retries/waits. One long one waiting for port 22 to be open, then a separate short one for trying the actual ssh connection. It's not uncommon to get stuck waiting forever (not literally, but 10 minutes feels like forever ;) ) on an ssh connection that will never work. That is all outside of the context of this issue though.
